packages/lisk-core/src/controller.js

```diff
@@ -1,40 +1,171 @@
-'use strict';
-
-const { ModuleFactory } = require('@lisk-core/utils');
-const Bus = require('./bus');
 const Promise = require('bluebird');
+const fs = require('fs-extra');
+const config = require('./helpers/config');
+const ModuleFactory = require('./factories/modules');
+const ComponentFactory = require('./factories/components');
+const EventEmitterChannel = require('./channels/event_emitter');
+const Schema = require('./helpers/schema');
+const Bus = require('./bus');
+
+const logger = ComponentFactory.create('logger', config.components.logger);
 
 module.exports = class Controller {
 	constructor(options) {
-		console.log('Initializing controller');
+		logger.info('Initializing controller');
+		this.config = options;
 		this.modules = {};
+		this.channel = null;
 
 		// Setting up bus
-		if(!this.bus) {
+		if (!this.bus) {
 			this.bus = new Bus(this, {
 				wildcard: true,
 				delimiter: ':',
-				maxListeners: 1000
+				maxListeners: 1000,
 			});
 		}
 	}
 
-	async loadModules() {
-		const chainModule = ModuleFactory.create('child_process', 'lisk-chain', {}, this, this.bus);
-		const p2p2Module = ModuleFactory.create('in_memory', 'lisk-p2p', {}, this, this.bus);
+	async setup() {
+		await Schema.setup();
+
+		const errors = Schema.validate(this.config, Schema.getSchema().config);
+		if (errors.length) {
+			logger.error(errors);
+			throw new Error('Configuration Validation Failed');
+		}
 
-		this.modules[chainModule.alias] = chainModule;
-		this.modules[p2p2Module.alias] = p2p2Module;
+		process.title = `Lisk ${this.config.pkg.version} : (${
+			this.config.dirs.root
+		})`;
 
-		await Promise.map(Object.keys(this.modules), async (m) => {
-			await this.modules[m].load();
-		});
+		// Make sure all directories exists
+		fs.emptyDirSync(this.config.dirs.temp);
+		Object.keys(this.config.dirs).forEach(dir =>
+			fs.ensureDirSync(this.config.dirs[dir]),
+		);
+		fs.writeFileSync(`${this.config.dirs.pids}/controller.pid`, process.pid);
+
+		// Set the custom directory to load modules
+		if (this.config.dirs.modules) {
+			logger.info(
+				`Setting custom module directory to: ${this.config.dirs.modules}`,
+			);
+		}
 
-		console.info(`Bus listening to events: ${this.bus.getEvents()}`);
-		console.info(`Bus ready for actions: ${this.bus.getActions()}`);
+		// Setup bus RPC socket
+		await this.bus.setup();
```

Comment:
bus setup can be moved to cli/index.js since it's being done once in application lifetime.